### 🖋 **Notebook Contents**

0. Initial Setup
1. Business Problem Understanding
2. Data Understanding

****

## `Initial Setup`

In [2]:
import pandas as pd

## `Business Problem Understanding`

**Context**:

Terdapat suatu perusahaan yang bergerak dibidang perbankan/keuangan, yaitu **Bank Beta**. Saat ini Bank Beta memiliki suatu data mengenai perilaku kurang lebih 9000 pengguna/nasabah yang menggunakan kartu kredit **selama 6 bulan kebelakang**. Beberapa perilaku tersebut antara lain, jumlah sisa saldo yang dapat digunakan (balance), seberapa sering nasabah menggunakan kartu kreditnya, jumlah saldo yang telah digunakan, dll.

**Problem Statement**:

Saat ini Marketing Departement masih kesulitan untuk mengadakan campaign marketing kepada para nasabahnya, karena masih belum mengetahui karakteristik dari para nasabahnya. Campaign marketing yang tepat sangat penting untuk dilakukan, berdasarkan referensi di bawah, jika melakukan campaign marketing yang tidak tepat sasaran dan nasabah memiliki pengalaman yang buruk (negative experiences) hanya satu kali saja, sehingga gagal untuk memenuhi ekspektasi dari nasabah, maka 61% dari total nasabah akan **pindah ke competitor lainnya**. Bahkan, akan meninggkat menjadi 76% jika nasabah memiliki pengalaman yang buruk (negative experiences) hingga berkali-kali. Jika hal ini terjadi pada Bank Beta, tentu hal ini akan berdampak buruk bagi revenue Bank Beta. Selain itu, rata-rata per tahun bank kehilangan nasabahnya sebesar 12.5%, sedangkan akuisisi nasabah baru sebesar 13%, namun biaya (cost) akuisisi nasabah baru 5 kali lebih besar daripada mempertahankan nasabah lama. Kemudian, jika ada pengurangan 5% (saja) dari nasabah yang hilang (customer turnover), maka bank bisa meningkatkan _net profit_ sebesar 80%. Oleh karena itu, Bank Beta ingin meminimalisir kemungkinan hal tersebut terjadi. Salah satu cara yang dapat dilakukan adalah dengan melakukan campaign marketing yang tepat ke nasabah yang tepat pula.

Tim Data Scientist akan merumuskan beberapa permasalahan, antara lain:
- Bagaimana pola-pola (pattern)/karakteristik tersembunyi dari para nasabah jika dilakukan pendekatan menggunakan algoritma _unsupervised machine larning_ ?
- Bagaimana karakteristik dari masing-masing cluster terbentuk ?
- .....

Referensi:
- [Effective data-driven campaigning for credit cards: target, attract, retain and engage](https://www.acuitykp.com/blog/effective-data-driven-campaigning-for-credit-cards/#:~:text=Industry%20studies%20suggest%20that%20if%20a%20provider%20fails%20to%20exceed%20customer%20expectations%2C%20after%20just%20one%20negative%20experience%2C%20more%20than%20half%20(61%25)%20of%20the%20customers%20say%20they%20would%20move%20to%20a%20competitor%3B%20this%20number%20rises%20to%2076%25%20in%20the%20case%20of%20multiple%20negative%20experiences.)
- [The Cost of Customer Churn](https://www.fpsc.com/the_cost_of_customer_churn.pdf)

**Goals**:

Bank Beta bisa mengetahui bagaimana karakteristik para nasabah. Dengan mengetahui karakteristik tersebut, tim Marketing Departement Bank Beta bisa melakukan campaign marketing yang tepat untuk para nasabahnya. Diharapkan dengan campaign marketing yang tepat dapat membuat para nasabah untuk semakin sering menggunakan kartu kreditnya, sehingga Bank beta dapat **meningkatkan _revenue_**.

.....

**Project Stakeholders**:

Marketing Departement: departement ini bertanggungjawab untuk memaksimalkan revenue dengan cara melakukan beberapa strategi marketing (misalnya, promosi atau diskon) kepada pelanggan, sehingga dengan model ini Marketing Departement bisa memberikan promosi/diskon yang tepat kepada pelanggan yang **tepat** pula.

......

**Analytic Approach**:

Tim Data Scientist akan melakukan beberapa untuk melakukan pendekatan untuk menjawab _Problem Statement_ di atas, antara lain:
1. **Visualisasi** data. Melihat bagaimana persebaran data berdasarkan visualisasi grafik yang terbentuk. Kemudian, mengambil _insight_ dari visualisasi tersebut. Metode ini merupakan salah satu metode yang cepat dan efektif untuk memahami suatu data pada tahap awal (initial judgement). Tim Data Scientist juga akan membuat _interactive dashboard_ menggunakan tools Tableau. .....
2. Analisa menggunakan metode **statistik**. Metode ini digunakan untuk mengvalidasi analisa secara terukur (measureable) menggunakan metode-metode statistik, seperti statistik deskriptif atau statistik inferensial. .....
3. **Machine Learning**. Untuk memperkuat metode 1 dan metode 2 diatas, jadi yang akan kami lakukan adalah menganalisa data untuk menemukan pola (pattern) suatu kesamaan/perbedaan karakteristik yang tersembunyi antar nasabah/pengguna (pengelompokkan/_clustering_) menggunakan algoritma _machine learning_. Kemudian kami akan membangun model **_unsupervised machine learning_ (clustering)** yang dapat membantu perusahaan untuk menemukan karakteristik-karateristik tersebut. Ketika ada nasabah baru, model ini nantinya akan berguna untuk memprediksi nasabah masuk ke dalam cluster yang mana, sehingga Marketing Departement bisa menentukan campaign marketing yang tepat.

**Metric Evaluation**:

Metrik evaluasi adalah ukuran kuantitatif yang digunakan untuk menilai kinerja dan efektivitas model machine learning. Metrik ini memberikan penjelasan tentang seberapa baik kinerja model dan dapat membantu dalam membandingkan berbagai model (benchmarking model). Pada data set ini juga tidak terdapat _ground truth label_, sehingga untuk pemilihan metrik evaluasi juga harus menggunakan metrik-metrik yang tidak menggunakan _ground truth label_.

Perlu diketahui bahwa basis metrik evaluasi pada _clustering_ umumnya adalah menggunakan jarak (distance) antar intra-cluster (self-cluster) dan inter-cluster (neighboring-cluster), di mana dikatakan bahwa _clustering_ yang bagus harus memiliki **jarak intra-cluster yang dekat** dan **jarak inter-cluster yang jauh**. Oleh karena itu, untuk evaluasi model unsupervised machine learning akan menggunakan metrik evaluasi silhouette score, Davies Bouldin Index (DBI), dan Calinski-Harabasz Index (CH).

- Silhouette score: mengukur jarak rata-rata antar satu sampel dengan sampel yang lainnya yang berada dalam satu cluster dan jarak rata-rata antar satu sampel dengan sampel yang lainnya yang berada pada cluster terdekat (next nearest cluster). Nilainya rentang antara -1 sampai 1, semakin mendekati 1, maka pengelompokkannya (clustering) semakin baik, atau dengan kata lain cluster satu dengan cluster yang lainnya terpisah secara baik (well separated).
- Davies Bouldin Index (DBI): konsepnya hampir sama dengan silhoutte score, yaitu menghitung jarak sampel, tetapi dalam hal ini, DBI memiliki konsep **SSW** (sum of square within cluster) dan **SSB** (sum of square between cluster). DBI mengukur jarak antar sampel dengan centroid cluster sampel tersebut dan mengukur jarak centroid antar clusternya. Score DBI yang mendekati 0 (nol), maka cluster-clusternya terpisah secara baik (well-separated).
- Calinski-Harabasz Index (CH): CH mengukur ratio dispersi antar cluster (SSB) dan intra-cluster (SSW). Tidak ada batasan nilai dama CH index, semakin besar nilainya, makanya pengelompokkanya (clustering) semakin bagus (well-separated).

References:
1. [scikit-learn clustering evaluation documentation](https://scikit-learn.org/stable/modules/clustering.html#clustering-evaluation)
2. [Three Performance Evaluation Metrics of Clustering When Ground Truth Labels Are Not Available](https://towardsdatascience.com/three-performance-evaluation-metrics-of-clustering-when-ground-truth-labels-are-not-available-ee08cb3ff4fb#:~:text=Main%20idea%20about%20clustering%20evaluation)
3. [Evaluasi Klustering Menggunakan Davies Bouldin Index (DBI) - youtube video LIA FAROKHAH channel](https://www.youtube.com/watch?v=zb8NkGDl5mk)

## `Data Understanding`

Dataset dapat diakses pada link berikut ini: [link](https://www.kaggle.com/datasets/arjunbhasin2013/ccdata)!

| Columns                                            | Definition                                                   | 
| ------------------------------------------------- | ------------------------------------------------------------ |
| `CUST_ID` | Identification of Credit Card holder (Categorical) |
| `BALANCE` | Balance amount left in their account to make purchases |
| `BALANCE_FREQUENCY` | How frequently the Balance is updated, score between 0 and 1 (1 = frequently updated, 0 = not frequently updated) |
| `PURCHASES` | Amount of purchases made from account |
| `ONEOFF_PURCHASES` | Maximum purchase amount done in one-go |
| `INSTALLMENTS_PURCHASES` | Amount of purchase done in installment |
| `CASH_ADVANCE` |  Cash in advance given by the user |
| `PURCHASES_FREQUENCY` | How frequently the Purchases are being made, score between 0 and 1 (1 = frequently purchased, 0 = not frequently purchased) |
| `ONEOFF_PURCHASES_FREQUENCY` | How frequently Purchases are happening in one-go (1 = frequently purchased, 0 = not frequently purchased) |
| `PURCHASES_INSTALLMENTS_FREQUENCY` | How frequently purchases in installments are being done (1 = frequently done, 0 = not frequently done) |
| `CASH_ADVANCE_FREQUENCY` | How frequently the cash in advance being paid |
| `CASH_ADVANCE_TRX` | Number of Transactions made with "Cash in Advanced" |
| `PURCHASES_TRX` | Number of purchase transactions made |
| `CREDIT_LIMIT` | Limit of Credit Card for user |
| `PAYMENTS` | Amount of Payment done by user |
| `MINIMUM_PAYMENTS` | Minimum amount of payments made by user |
| `PRC_FULL_PAYMENT` | Percent of full payment paid by user |
| `TENURE` | Tenure of credit card service for user |

In [3]:
data = pd.read_csv("../data/raw/CC GENERAL.csv")
data

CUST_ID      BALANCE  BALANCE_FREQUENCY  PURCHASES  ONEOFF_PURCHASES  \
0     C10001    40.900749           0.818182      95.40              0.00   
1     C10002  3202.467416           0.909091       0.00              0.00   
2     C10003  2495.148862           1.000000     773.17            773.17   
3     C10004  1666.670542           0.636364    1499.00           1499.00   
4     C10005   817.714335           1.000000      16.00             16.00   
...      ...          ...                ...        ...               ...   
8945  C19186    28.493517           1.000000     291.12              0.00   
8946  C19187    19.183215           1.000000     300.00              0.00   
8947  C19188    23.398673           0.833333     144.40              0.00   
8948  C19189    13.457564           0.833333       0.00              0.00   
8949  C19190   372.708075           0.666667    1093.25           1093.25   

      INSTALLMENTS_PURCHASES  CASH_ADVANCE  PURCHASES_FREQUENCY  \
0                      95.40      0.000000             0.166667   
1                       0.00   6442.945483             0.000000   
2                       0.00      0.000000             1.000000   
3                       0.00    205.788017             0.083333   
4                       0.00      0.000000             0.083333   
...                      ...           ...                  ...   
8945                  291.12      0.000000             1.000000   
8946                  300.00      0.000000             1.000000   
8947                  144.40      0.000000             0.833333   
8948                    0.00     36.558778             0.000000   
8949                    0.00    127.040008             0.666667   

      ONEOFF_PURCHASES_FREQUENCY  PURCHASES_INSTALLMENTS_FREQUENCY  \
0                       0.000000                          0.083333   
1                       0.000000                          0.000000   
2                       1.000000                          0.000000   
3                       0.083333                          0.000000   
4                       0.083333                          0.000000   
...                          ...                               ...   
8945                    0.000000                          0.833333   
8946                    0.000000                          0.833333   
8947                    0.000000                          0.666667   
8948                    0.000000                          0.000000   
8949                    0.666667                          0.000000   

      CASH_ADVANCE_FREQUENCY  CASH_ADVANCE_TRX  PURCHASES_TRX  CREDIT_LIMIT  \
0                   0.000000                 0              2        1000.0   
1                   0.250000                 4              0        7000.0   
2                   0.000000                 0             12        7500.0   
3                   0.083333                 1              1        7500.0   
4                   0.000000                 0              1        1200.0   
...                      ...               ...            ...           ...   
8945                0.000000                 0              6        1000.0   
8946                0.000000                 0              6        1000.0   
8947                0.000000                 0              5        1000.0   
8948                0.166667                 2              0         500.0   
8949                0.333333                 2             23        1200.0   

         PAYMENTS  MINIMUM_PAYMENTS  PRC_FULL_PAYMENT  TENURE  
0      201.802084        139.509787          0.000000      12  
1     4103.032597       1072.340217          0.222222      12  
2      622.066742        627.284787          0.000000      12  
3        0.000000               NaN          0.000000      12  
4      678.334763        244.791237          0.000000      12  
...           ...               ...               ...     ...  
8945   325.594462         48.886365  

In [4]:
display(
    # numerical variable
    data.describe(),
    # categorical variable
    data.describe(include='object')
)

BALANCE  BALANCE_FREQUENCY     PURCHASES  ONEOFF_PURCHASES  \
count   8950.000000        8950.000000   8950.000000       8950.000000   
mean    1564.474828           0.877271   1003.204834        592.437371   
std     2081.531879           0.236904   2136.634782       1659.887917   
min        0.000000           0.000000      0.000000          0.000000   
25%      128.281915           0.888889     39.635000          0.000000   
50%      873.385231           1.000000    361.280000         38.000000   
75%     2054.140036           1.000000   1110.130000        577.405000   
max    19043.138560           1.000000  49039.570000      40761.250000   

       INSTALLMENTS_PURCHASES  CASH_ADVANCE  PURCHASES_FREQUENCY  \
count             8950.000000   8950.000000          8950.000000   
mean               411.067645    978.871112             0.490351   
std                904.338115   2097.163877             0.401371   
min                  0.000000      0.000000             0.000000   
25%                  0.000000      0.000000             0.083333   
50%                 89.000000      0.000000             0.500000   
75%                468.637500   1113.821139             0.916667   
max              22500.000000  47137.211760             1.000000   

       ONEOFF_PURCHASES_FREQUENCY  PURCHASES_INSTALLMENTS_FREQUENCY  \
count                 8950.000000                       8950.000000   
mean                     0.202458                          0.364437   
std                      0.298336                          0.397448   
min                      0.000000                          0.000000   
25%                      0.000000                          0.000000   
50%                      0.083333                          0.166667   
75%                      0.300000                          0.750000   
max                      1.000000                          1.000000   

       CASH_ADVANCE_FREQUENCY  CASH_ADVANCE_TRX  PURCHASES_TRX  CREDIT_LIMIT  \
count             8950.000000       8950.000000    8950.000000   8949.000000   
mean                 0.135144          3.248827      14.709832   4494.449450   
std                  0.200121          6.824647      24.857649   3638.815725   
min                  0.000000          0.000000       0.000000     50.000000   
25%                  0.000000          0.000000       1.000000   1600.000000   
50%                  0.000000          0.000000       7.000000   3000.000000   
75%                  0.222222          4.000000      17.000000   6500.000000   
max                  1.500000        123.000000     358.000000  30000.000000   

           PAYMENTS  MINIMUM_PAYMENTS  PRC_FULL_PAYMENT       TENURE  
count   8950.000000       8637.000000       8950.000000  8950.000000  
mean    1733.143852        864.206542          0.153715    11.517318  
std     2895.063757       2372.446607          0.292499     1.338331  
min        0.000000          0.019163          0.000000     6.000000  
25%      383.276166        169.123707          0.000000    12.000000  
50%      856.901546        312.343947          0.000000    12.000000  
75%     1901.134317        825.485459          0.142857    12.000000  
max    50721.483360      76406.207520          1.000000    12.000000

CUST_ID
count     8950
unique    8950
top     C10001
freq         1

**_Insight_**:
- Nilai mininum dan maksimum pada semua kolom bervariasi. Namun, mayoritas nilai minimumnya adalah 0
- Perlu melakukan re-scalling karena pada dasarnya algoritma clustering berbasis "jarak"
- Terdapat 8950 nasabah
- .....
- dst.

In [5]:
# jumlah baris dan kolom
print(f'Jumlah baris dan kolom: {data.shape[0]} baris dan {data.shape[1]} kolom')

# skimming dataset
pd.DataFrame(
                {
                'kolom': data.columns.values,
                'data_type': data.dtypes.values,
                'null_value(%)': data.isna().mean().values * 100,
                'n_unique': data.nunique().values,
                'zero_value' : [True if (data[col] == 0).any() else False for col in data.columns],
                'neg_value' : [True if (data[col].dtype == int or data[col].dtype == float) and (data[col] < 0).any() else False for col in data.columns],
                'min': data.min().values,
                'max': data.max().values,
                'sample_unique': [data[col].unique() for col in data.columns]
                }
            )

Jumlah baris dan kolom: 8950 baris dan 18 kolom


kolom data_type  null_value(%)  n_unique  \
0                            CUST_ID    object       0.000000      8950   
1                            BALANCE   float64       0.000000      8871   
2                  BALANCE_FREQUENCY   float64       0.000000        43   
3                          PURCHASES   float64       0.000000      6203   
4                   ONEOFF_PURCHASES   float64       0.000000      4014   
5             INSTALLMENTS_PURCHASES   float64       0.000000      4452   
6                       CASH_ADVANCE   float64       0.000000      4323   
7                PURCHASES_FREQUENCY   float64       0.000000        47   
8         ONEOFF_PURCHASES_FREQUENCY   float64       0.000000        47   
9   PURCHASES_INSTALLMENTS_FREQUENCY   float64       0.000000        47   
10            CASH_ADVANCE_FREQUENCY   float64       0.000000        54   
11                  CASH_ADVANCE_TRX     int64       0.000000        65   
12                     PURCHASES_TRX     int64       0.000000       173   
13                      CREDIT_LIMIT   float64       0.011173       205   
14                          PAYMENTS   float64       0.000000      8711   
15                  MINIMUM_PAYMENTS   float64       3.497207      8636   
16                  PRC_FULL_PAYMENT   float64       0.000000        47   
17                            TENURE     int64       0.000000         7   

    zero_value  neg_value       min          max  \
0        False      False    C10001       C19190   
1         True      False       0.0  19043.13856   
2         True      False       0.0          1.0   
3         True      False       0.0     49039.57   
4         True      False       0.0     40761.25   
5         True      False       0.0      22500.0   
6         True      False       0.0  47137.21176   
7         True      False       0.0          1.0   
8         True      False       0.0          1.0   
9         True      False       0.0          1.0   
10        True      False       0.0          1.5   
11        True      False         0          123   
12        True      False         0          358   
13       False      False      50.0      30000.0   
14        True      False       0.0  50721.48336   
15       False      False  0.019163  76406.20752   
16        True      False       0.0          1.0   
17       False      False         6           12   

                                        sample_unique  
0   [C10001, C10002, C10003, C10004, C10005, C1000...  
1   [40.900749, 3202.467416, 2495.148862, 1666.670...  
2   [0.818182, 0.909091, 1.0, 0.636364, 0.545455, ...  
3   [95.4, 0.0, 773.17, 1499.0, 16.0, 1333.28, 709...  
4   [0.0, 773.17, 1499.0, 16.0, 6402.63, 661.49, 1...  
5   [95.4, 0.0, 1333.28, 688.38, 436.2, 200.0, 920...  
6   [0.0, 6442.945483, 205.788017, 346.81139, 2301...  
7   [0.166667, 0.0, 1.0, 0.083333, 0.666667, 0.333...  
8   [0.0, 1.0, 0.083333, 0.166667, 0.25, 0.916667,...  
9   [0.083333, 0.0, 0.583333, 1.0, 0.25, 0.916667,...  
10  [0.0, 0.25, 0.083333, 0.166667, 0.333333, 0.36...  
11  [0, 4, 1, 3, 6, 13, 5, 16, 10, 2, 11, 7, 12, 3...  
12  [2, 0, 12, 1, 8, 64, 5, 3, 6, 26, 11, 9, 92, 1...  
13  [1000.0, 7000.0, 7500.0, 1200.0, 1800.0, 13500...  
14  [201.802084, 4103.032597, 622.066742, 0.0, 678...  
15  [139.509787, 1072.340217, 627.284787, nan, 244...  
16  [0.0, 0.222222, 1.0, 0.25, 0.083333, 0.3, 0.33...  
17                           [12, 8, 11, 9, 10, 7, 6]

**_Insight_**:
- Dataset terdiri dari **8950 baris dan 18 kolom**. Setiap baris data menunjukkan riwayat transaksi satu nasabah Bank Beta. Riwayat transaksi tersebut meliputi antara lain, sisa saldo kartu kredit yang dapat digunakan, seberapa sering nasabah meng-update saldo kartu kredit, seberapa sering nasabah melakukan pembelian dengan kartu kredit, jumlah pembelian yang telah dilakukan baik yang dibayar secara lunas maupun angsuran, limit  dll. Berdasarkan data-data di atas akan dilakukan pengecekan lebih lanjut apakah data-data tersebut sudah cukup untuk melakukan pengelompokkan nasabah berdasarkan karakteristiknya.
- Jenis data:
    - Numerik:
        - Continue: BALANCE, PURCHASES, ONEOFF_PURCHASES, INSTALLMENTS_PURCHASES, CASH_ADVANCE, CREDIT_LIMIT, PAYMENTS, MINIMUM_PAYMENTS
        - Diskrit: CASH_ADVANCE_TRX, PURCHASES_TRX
        - Ordinal: BALANCE_FREQUENCY, PURCHASES_FREQUENCY, ONEOFF_PURCHASES_FREQUENCY, PURCHASES_INSTALLMENTS_FREQUENCY, CASH_ADVANCE_FREQUENCY, PRC_FULL_PAYMENT, TENURE
    - Kategorikal:
        - Nominal: CUST_ID
        - Ordinal: -
- Tipe data pada semua kolom sudah sesuai
- Terdapat _missing values_ sebanyak 0.01% pada kolom CREDIT_LIMIT dan 3.49% pada kolom MINIMUM PAYMENTS dari total data yang tersedia
- Tidak terdapat nilai negatif pada semua kolom
- .....
- .....
- dst.

Note: untuk kolom yang bersifat data numerik ordinal, nilai pada kolom-kolom tersebut memiliki tipe data interger atau float, namun juga memiliki urutan/tingkatan/kasta. [(referensi)](https://www.dummies.com/article/academics-the-arts/math/statistics/types-of-statistical-data-numerical-categorical-and-ordinal-169735/#:~:text=Ordinal%20data%20mixes%20numerical%20and%20categorical%20data.%20The%20data%20fall%20into%20categories%2C%20but%20the%20numbers%20placed%20on%20the%20categories%20have%20meaning.%20For%20example%2C%20rating%20a%20restaurant%20on%20a%20scale%20from%200%20(lowest)%20to%204%20(highest)%20stars%20gives%20ordinal%20data.)

**Apakah data sudah cukup untuk menjawab problem statement dan goal yang ingin dicapai ?**

Beberapa sumber di bawah menyebutkan bahwa untuk dapat melakukan segmentasi kartu kredit, kita perlu mempunyai beberapa data para nasabah. **Perusahaan Mckinsey** menyebutkan bahwa secara garis besar data yang dibutuhkan adalah perilaku (behavior) dan demografi nasabah. Data perilaku nasabah dapat meliputi riwayat transaksi nasabah, sedangkan data demografi nasabah dapat meliputi umur nasabah, status pernikahan, domisili, dll. Kemudian, jika berdasarkan tulisan artikel **Sadrach Pierre, Ph.D**, senior data scientist WorldQuant Predictive, segmentasi dapat dilakukan berdasarkan informasi sebarapa sering nasabah melakukan pembelian dengan kartu kredit, jumlah uang yang telah digunakan, tanggal transaksi, dsb. Oleh karena itu, tim Data Scientist menyimpulkan bahwa data para nasabah yang telah dimiliki oleh Bank Beta **sudah cukup** menjawab problem statement dan mencapai goal yang ingin dicapai.

Referensi:
- [New frontiers in credit card segmentation: Tapping unmet consumer needs](https://www.mckinsey.com/~/media/mckinsey/dotcom/client_service/financial%20services/latest%20thinking/payments/mop19_new%20frontiers%20in%20credit%20card%20segmentation.ashx)
- [Mastering Customer Segmentation using Credit Card Transaction Data](https://towardsdatascience.com/mastering-customer-segmentation-using-credit-card-transaction-data-dc39a8465766#:~:text=Segments%20can%20be%20created%20using%20information%20like%20frequency%20of%20purchases%2C%20transaction%20amounts%2C%20purchase%20dates%20and%20more.%20All%20of%20these%20characteristics%20can%20be%20used%20to%20generate%20well%20defined%20clusters%20with%20easy%20to%20interpret%20characteristics.)

--------------
--------------
--------------